tests/unit/s2n_ktls_io_test.c

```diff
@@ -286,5 +308,212 @@ int main(int argc, char **argv)
+            size_t bytes_read = 0;
+            /* recv should block since there is no data */
+            for (size_t i = 0; i < blocked_invoked_count; i++) {
+                EXPECT_ERROR_WITH_ERRNO(
+                        s2n_ktls_recvmsg(client, &recv_record_type, recv_buf, S2N_TEST_TO_SEND, &blocked, &bytes_read),
+                        S2N_ERR_IO_BLOCKED);
+                EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+            }
+
+            /* send data to unblock */
+            struct iovec msg_iov = { .iov_base = test_data, .iov_len = S2N_TEST_TO_SEND };
+            size_t bytes_written = 0;
+            EXPECT_OK(s2n_ktls_sendmsg(server, test_record_type, &msg_iov, 1, &blocked, &bytes_written));
+            EXPECT_EQUAL(bytes_written, S2N_TEST_TO_SEND);
+
+            EXPECT_OK(s2n_ktls_recvmsg(client, &recv_record_type, recv_buf, S2N_TEST_TO_SEND, &blocked, &bytes_read));
+            EXPECT_BYTEARRAY_EQUAL(test_data, recv_buf, bytes_read);
+            EXPECT_EQUAL(bytes_read, bytes_written);
+
+            /* recv should block again since we have read all the data */
+            for (size_t i = 0; i < blocked_invoked_count; i++) {
+                EXPECT_ERROR_WITH_ERRNO(
+                        s2n_ktls_recvmsg(client, &recv_record_type, recv_buf, S2N_TEST_TO_SEND, &blocked, &bytes_read),
+                        S2N_ERR_IO_BLOCKED);
+                EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+            }
+
+            EXPECT_EQUAL(io_pair.client_in.recvmsg_invoked_count, (blocked_invoked_count * 2) + 1);
+            EXPECT_EQUAL(io_pair.client_in.sendmsg_invoked_count, 1);
+        };
+
+        /* Both EWOULDBLOCK and EAGAIN should return a S2N_ERR_IO_BLOCKED error */
+        {
+            DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                    s2n_connection_ptr_free);
+            struct s2n_test_ktls_io_fail_ctx io_ctx = { 0 };
+            EXPECT_OK(s2n_ktls_set_recvmsg_cb(client, s2n_test_ktls_recvmsg_fail, &io_ctx));
+
+            uint8_t recv_buf[S2N_TLS_MAXIMUM_FRAGMENT_LENGTH] = { 0 };
+            s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+            uint8_t recv_record_type = 0;
+            size_t bytes_read = 0;
+
+            io_ctx.errno_code = EWOULDBLOCK;
+            EXPECT_ERROR_WITH_ERRNO(
+                    s2n_ktls_recvmsg(client, &recv_record_type, recv_buf, S2N_TEST_TO_SEND, &blocked, &bytes_read),
+                    S2N_ERR_IO_BLOCKED);
+            EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+
+            /* cppcheck-suppress redundantAssignment */
+            io_ctx.errno_code = EAGAIN;
+            EXPECT_ERROR_WITH_ERRNO(
+                    s2n_ktls_recvmsg(client, &recv_record_type, recv_buf, S2N_TEST_TO_SEND, &blocked, &bytes_read),
+                    S2N_ERR_IO_BLOCKED);
+            EXPECT_EQUAL(blocked, S2N_BLOCKED_ON_READ);
+
+            EXPECT_EQUAL(io_ctx.invoked_count, 2);
```

Comment:
What's the difference between the io_ctx.invoked_count and the sendmsg/recvmsg invoked count? Why check one over the other?

Reply:
They are simply different contexts. The `s2n_test_ktls_io_stuffer` has the `sendmsg/recvmsg_invoked_count` since its more complicated.

This one tests the more simple failure case so uses  `s2n_test_ktls_io_fail_ctx` and only has the `invoked_count` field.